overlord/ifacestate/hotplug.go

```diff
@@ -30,3 +35,98 @@ func (m *InterfaceManager) HotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo
 // HotplugDeviceRemoved gets called when a device is removed from the system.
 func (m *InterfaceManager) HotplugDeviceRemoved(devinfo *hotplug.HotplugDeviceInfo) {
 }
+
+// ensureUniqueName modifies proposedName so that it's unique according to isUnique predicate.
+// Uniqueness is achieved by appending a numeric suffix, or increasing existing suffix.
+func ensureUniqueName(proposedName string, isUnique func(string) bool) string {
+	// if the name is unique right away, do nothing
+	if isUnique(proposedName) {
+		return proposedName
+	}
+
+	suffixNumValue := 0
+	prefix := strings.TrimRightFunc(proposedName, unicode.IsDigit)
+	if prefix != proposedName {
+		suffixNumValue, _ = strconv.Atoi(proposedName[len(prefix):])
+	}
+	prefix = strings.TrimRight(prefix, "-")
+
+	// increase suffix value until we have a unique name
+	for {
+		suffixNumValue++
+		proposedName = fmt.Sprintf("%s%d", prefix, suffixNumValue)
+		if isUnique(proposedName) {
+			return proposedName
+		}
+	}
+}
+
+const maxGenerateSlotNameLen = 20
+
+// makeSlotName sanitizes a string to make it a valid slot name that
+// passes validation rules implemented by ValidateSlotName (see snap/validate.go):
+// - only lowercase letter, digits and dashes are allowed
+// - must start with a letter
+// - no double dashes, cannot end with a dash.
+// In addition names are truncated not to exceed maxGenerateSlotNameLen characters.
+func makeSlotName(s string) string {
+	var out []rune
+	var charCount int
+	// the dash flag is used to prevent consecutive dashes, and the dash in the front
+	dash := true
+Loop:
+	for _, c := range s {
+		switch {
+		case c == '-' && !dash:
+			dash = true
+			out = append(out, '-')
+		case unicode.IsLetter(c):
+			out = append(out, unicode.ToLower(c))
+			dash = false
+			charCount++
+			if charCount >= maxGenerateSlotNameLen {
+				break Loop
+			}
+		case unicode.IsDigit(c) && charCount > 0:
+			out = append(out, c)
+			dash = false
+			charCount++
+			if charCount >= maxGenerateSlotNameLen {
+				break Loop
+			}
+		default:
+			// any other character is ignored
+		}
+	}
+	// make sure the name doesn't end with a dash
+	return strings.TrimRight(string(out), "-")
+}
+
+var nameAttrs = []string{"NAME", "ID_MODEL_FROM_DATABASE", "ID_MODEL"}
+
+// suggestedSlotName returns the shortest name derived from attributes defined
+// by nameAttrs, or the fallbackName if there is no known attribute to derive
+// name from. The name created from attributes is sanitized to ensure it's a
+// valid slot name. The fallbackName is typically the name of the interface.
+func suggestedSlotName(devinfo *hotplug.HotplugDeviceInfo, fallbackName string) string {
+	var candidates []string
+	for _, attr := range nameAttrs {
+		name, ok := devinfo.Attribute(attr)
+		if ok {
+			name = makeSlotName(name)
+			if name != "" {
+				candidates = append(candidates, name)
```

Comment:
Why is this building a list of names to then compute the shortest one, instead of simply comparing len(name) with len(current candidate) here?

Reply:
Good point, thanks, simplified.